### PR TITLE
[Feature] support join on complex types

### PR DESCRIPTION
**Pull request by fzhedu**
1. support join on array/map/struct types.
2. disable non-equal binary operators on complex types
3. fix a bug where reporting casting error
4. fix a bug where binary predict of arrays don't process const inputs

```
mysql> select s.i_1, t.i_1 from array_test s join array_test t on s.d_1 = t.d_1;
+--------------+--------------+
| i_1          | i_1          |
+--------------+--------------+
| NULL         | NULL         |
| [1,2,3,4,10] | [1,2,3,4,10] |
| [1,2]        | [1,2]        |
| [1,null]     | [1,null]     |
| [1,2,3,4,10] | [1,2,3,4,10] |
| [1,2,3,4,10] | [1,2,3,4,10] |
+--------------+--------------+
6 rows in set (0.04 sec)

mysql> select t.s1, s.s4 from struct_test t join struct_test s where s.s1 <=> t.s4 order by t.pk;
+-----------------------+-----------------------+
| s1                    | s4                    |
+-----------------------+-----------------------+
| {"c0":null,"c1":null} | {"c0":null,"c1":null} |
| NULL                  | NULL                  |
+-----------------------+-----------------------+
2 rows in set (0.02 sec)

mysql> select map0 from map_test t where exists (select 1 from map_test s where t.map0 = s.map0);
+------------------+
| map0             |
+------------------+
| {1:null}         |
| {null:null}      |
| {0:"ab"}         |
| {3:"",null:null} |
+------------------+
4 rows in set (0.03 sec)
```
## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

File: be/src/runtime/types.cpp [hunk 1]
```diff
@@ -261,16 +261,46 @@ std::string TypeDescriptor::debug_string() const {
...
-    return type != TYPE_JSON && type != TYPE_OBJECT && type != TYPE_PERCENTILE && type != TYPE_HLL &&
-           type != TYPE_MAP && type != TYPE_STRUCT && type != TYPE_ARRAY;
+    if (type == TYPE_ARRAY) {
+        return children[0].support_join();
+    }
+    if (type == TYPE_MAP) {
+        return children[0].support_join() && children[1].support_join();
+    }
+    if (type == TYPE_STRUCT) {
+        for (auto& c : children) {
```
Comment (silverbullet233) [inline be/src/runtime/types.cpp:271]: suggest to use std::all_of to simplify code

File: be/src/runtime/types.cpp [hunk 2]
```diff
@@ -261,16 +261,46 @@ std::string TypeDescriptor::debug_string() const {
...
+    if (type == TYPE_MAP) {
+        return children[0].support_groupby() && children[1].support_groupby();
+    }
+    if (type == TYPE_STRUCT) {
+        for (auto& c : children) {
+            if (!c.support_groupby()) {
+                return false;
+            }
+        }
+        return true;
```
Comment (silverbullet233) [inline be/src/runtime/types.cpp:302]: similar as above

File: fe/fe-core/src/main/java/com/starrocks/analysis/BinaryPredicate.java [hunk 3]
```diff
@@ -192,16 +193,8 @@ public static Type getCmpType(Type type1, Type type2) {
...
-            // Must both be array
-            if (!type1.matchesType(type2)) {
-                return Type.INVALID;
-            }
-            return type1;
-        }
         if (type1.isComplexType() || type2.isComplexType()) {
-            // We don't support complex type for binary predicate.
-            return Type.INVALID;
+            return TypeManager.getCommonSuperType(type1, type2);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/analysis/BinaryPredicate.java:197]: can't modify here, we only support equals, the method will check `><` too

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 16 / 22 (72.73%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|0|2|00.00%|[147, 157]|
|:large_blue_circle:|com/starrocks/analysis/BinaryPredicate.java|0|1|00.00%|[197]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|16|19|84.21%|[746, 751, 770]|

File: be/src/runtime/types.cpp (hunk 1 above)
Comment (fzhedu) [reply be/src/runtime/types.cpp]: changed

File: be/src/runtime/types.cpp (hunk 2 above)
Comment (fzhedu) [reply be/src/runtime/types.cpp]: changed too

File: fe/fe-core/src/main/java/com/starrocks/analysis/BinaryPredicate.java (hunk 3 above)
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/analysis/BinaryPredicate.java]: disable the function

File: be/src/runtime/types.cpp [hunk 4]
```diff
@@ -261,16 +261,24 @@ std::string TypeDescriptor::debug_string() const {
...
+    if (type == TYPE_ARRAY) {
+        return children[0].support_orderby();
+    }
     return type != TYPE_JSON && type != TYPE_OBJECT && type != TYPE_PERCENTILE && type != TYPE_HLL &&
            type != TYPE_MAP && type != TYPE_STRUCT;
 }
 
 bool TypeDescriptor::support_groupby() const {
+    if (type == TYPE_ARRAY || type == TYPE_MAP || type == TYPE_STRUCT) {
+        return std::all_of(children.begin(), children.end(), [](const TypeDescriptor& t) { return t.support_join(); });
```
Comment (silverbullet233) [inline be/src/runtime/types.cpp:280]: support_groupby？

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 16 / 21 (76.19%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|0|2|00.00%|[147, 157]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|16|19|84.21%|[746, 751, 770]|

File: be/src/runtime/types.cpp (hunk 4 above)
Comment (fzhedu) [reply be/src/runtime/types.cpp:280]: yes, in preview PR

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 5]
```diff
@@ -807,7 +832,7 @@ public boolean supportBloomFilter() {
     }
 
     public static final String NOT_SUPPORT_JOIN_ERROR_MSG =
-            "Type (nested) percentile/hll/bitmap/json/struct/map not support join";
+            "Type (nested) percentile/hll/bitmap/json not support join";
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:835]: why not support json btw

**Review by murphyatwork — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java (hunk 5 above)
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:835]: as the serialized results of json columns are not stable @Seaven

File: be/src/runtime/types.cpp (hunk 4 above)
Comment (fzhedu) [reply be/src/runtime/types.cpp]: fixed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 16 / 22 (72.73%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|0|2|00.00%|[147, 157]|
|:large_blue_circle:|com/starrocks/analysis/BinaryPredicate.java|0|1|00.00%|[197]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|16|19|84.21%|[746, 751, 770]|

**Review by stdpain — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/analysis/BinaryPredicate.java (hunk 3 above)
Comment (stdpain) [reply fe/fe-core/src/main/java/com/starrocks/analysis/BinaryPredicate.java:197]: Is it expected? maybe we shouldn't change this function

Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/analysis/BinaryPredicate.java:197]: just allow  array type to do any binary predicts, map/struct only do equal binary predicts.

**Review by murphyatwork — APPROVED** [review]

**Comment by fzhedu** [comment]
@mergify backport branch-3.1

**Pull request merged**